File: src/filter_ops.py
import cv2
import numpy as np

class FilterOps:
    @staticmethod
    def blur(frames, kernel):
        return [cv2.GaussianBlur(f, (kernel, kernel), 0) for f in frames]
    
    @staticmethod
    def sharpen(frames):
        k = np.array([[-1,-1,-1],[-1,9,-1],[-1,-1,-1]])
        return [cv2.filter2D(f, -1, k) for f in frames]
    
    @staticmethod
    def edge(frames):
        result = []
        for f in frames:
            e = cv2.Canny(f, 100, 200)
            result.append(cv2.cvtColor(e, cv2.COLOR_GRAY2BGR))
        return result
    
    @staticmethod
    def denoise(frames):
        return [cv2.fastNlMeansDenoisingColored(f, None, 10, 10, 7, 21) for f in frames]
    
    @staticmethod
    def threshold(frames, thresh):
        result = []
        for f in frames:
            g = cv2.cvtColor(f, cv2.COLOR_BGR2GRAY)
            _, th = cv2.threshold(g, thresh, 255, cv2.THRESH_BINARY)
            result.append(cv2.cvtColor(th, cv2.COLOR_GRAY2BGR))
        return result
    
    @staticmethod
    def emboss(frames):
        k = np.array([[-2, -1, 0], [-1, 1, 1], [0, 1, 2]])
        result = []
        for f in frames:
            gray = cv2.cvtColor(f, cv2.COLOR_BGR2GRAY)
            embossed = cv2.filter2D(gray, -1, k)
            result.append(cv2.cvtColor(embossed, cv2.COLOR_GRAY2BGR))
        return result
    
    @staticmethod
    def cartoon(frames):
        result = []
        for f in frames:
            gray = cv2.cvtColor(f, cv2.COLOR_BGR2GRAY)
            gray = cv2.medianBlur(gray, 5)
            edges = cv2.adaptiveThreshold(gray, 255, cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY, 9, 9)
            color = cv2.bilateralFilter(f, 9, 300, 300)
            cartoon = cv2.bitwise_and(color, color, mask=edges)
            result.append(cartoon)
        return result
    
    @staticmethod
    def vignette(frames, intensity):
        result = []
        for f in frames:
            h, w = f.shape[:2]
            kernel_x = cv2.getGaussianKernel(w, w * intensity)
            kernel_y = cv2.getGaussianKernel(h, h * intensity)
            kernel = kernel_y * kernel_x.T
            mask = kernel / kernel.max()
            mask = cv2.cvtColor(mask, cv2.COLOR_GRAY2BGR)
            result.append((f * mask).astype(np.uint8))
        return result
    
    @staticmethod
    def motion_blur(frames, size, angle):
        k = np.zeros((size, size))
        k[int((size-1)/2), :] = np.ones(size)
        k = cv2.warpAffine(k, cv2.getRotationMatrix2D((size/2-0.5, size/2-0.5), angle, 1.0), (size, size))
        k = k * (1.0 / np.sum(k))
        return [cv2.filter2D(f, -1, k) for f in frames]

File: src/FilterOps.cpp
#include "../include/FilterOps.h"

using namespace cv;

std::vector<Mat> FilterOps::blur(const std::vector<Mat>& frames, int kernel) {
    std::vector<Mat> result;
    for (const auto& f : frames) { Mat blurred; GaussianBlur(f, blurred, Size(kernel, kernel), 0); result.push_back(blurred); }
    return result;
}

std::vector<Mat> FilterOps::sharpen(const std::vector<Mat>& frames) {
    std::vector<Mat> result; Mat k = (Mat_<float>(3,3) << -1,-1,-1,-1,9,-1,-1,-1,-1);
    for (const auto& f : frames) { Mat sharp; filter2D(f, sharp, -1, k); result.push_back(sharp); }
    return result;
}

std::vector<Mat> FilterOps::edge(const std::vector<Mat>& frames) {
    std::vector<Mat> result; for (const auto& f : frames) { Mat e; Canny(f, e, 100, 200); Mat bgr; cvtColor(e, bgr, COLOR_GRAY2BGR); result.push_back(bgr); } return result;
}

std::vector<Mat> FilterOps::denoise(const std::vector<Mat>& frames) { std::vector<Mat> result; for (const auto& f : frames) { Mat denoised; fastNlMeansDenoisingColored(f, denoised, 10, 10, 7, 21); result.push_back(denoised); } return result; }

std::vector<Mat> FilterOps::threshold(const std::vector<Mat>& frames, int thresh) { std::vector<Mat> result; for (const auto& f : frames) { Mat g; cvtColor(f, g, COLOR_BGR2GRAY); Mat th; cv::threshold(g, th, thresh, 255, THRESH_BINARY); Mat bgr; cvtColor(th, bgr, COLOR_GRAY2BGR); result.push_back(bgr); } return result; }

std::vector<Mat> FilterOps::emboss(const std::vector<Mat>& frames) { std::vector<Mat> result; Mat k = (Mat_<float>(3,3) << -2,-1,0,-1,1,1,0,1,2); for (const auto& f : frames) { Mat gray; cvtColor(f, gray, COLOR_BGR2GRAY); Mat embossed; filter2D(gray, embossed, -1, k); Mat bgr; cvtColor(embossed, bgr, COLOR_GRAY2BGR); result.push_back(bgr); } return result; }

std::vector<Mat> FilterOps::cartoon(const std::vector<Mat>& frames) { std::vector<Mat> result; for (const auto& f : frames) { Mat gray; cvtColor(f, gray, COLOR_BGR2GRAY); medianBlur(gray, gray, 5); Mat edges; adaptiveThreshold(gray, edges, 255, ADAPTIVE_THRESH_MEAN_C, THRESH_BINARY, 9, 9); Mat color; bilateralFilter(f, color, 9, 300, 300); Mat cartoon; color.copyTo(cartoon, edges); result.push_back(cartoon); } return result; }

std::vector<Mat> FilterOps::vignette(const std::vector<Mat>& frames, double intensity) { std::vector<Mat> result; for (const auto& f : frames) { int h = f.rows, w = f.cols; Mat kernelX = getGaussianKernel(w, w * intensity, CV_32F); Mat kernelY = getGaussianKernel(h, h * intensity, CV_32F); Mat kernel = kernelY * kernelX.t(); double minVal, maxVal; minMaxLoc(kernel, &minVal, &maxVal); kernel = kernel / maxVal; Mat mask; cvtColor(kernel, mask, COLOR_GRAY2BGR); Mat vignetted; f.convertTo(vignetted, CV_32F); multiply(vignetted, mask, vignetted); vignetted = min(vignetted, 255.0); vignetted = max(vignetted, 0.0); vignetted.convertTo(vignetted, CV_8U); result.push_back(vignetted); } return result; }

std::vector<Mat> FilterOps::motionBlur(const std::vector<Mat>& frames, int size, double angle) { std::vector<Mat> result; Mat k = Mat::zeros(size, size, CV_32F); int mid = (size - 1) / 2; for (int i = 0; i < size; i++) k.at<float>(mid, i) = 1.0f; Point2f center(size / 2.0f - 0.5f, size / 2.0f - 0.5f); Mat M = getRotationMatrix2D(center, angle, 1.0); warpAffine(k, k, M, Size(size, size)); cv::Scalar sumVal = cv::sum(k); k = k / sumVal[0]; for (const auto& f : frames) { Mat blurred; filter2D(f, blurred, -1, k); result.push_back(blurred); } return result; }
